rules/ast/functions.go

```diff
@@ -406,6 +413,49 @@ func absentImpl(timestamp clientmodel.Timestamp, args []Node) interface{} {
 	}
 }
 
+// === deriv(node MatrixNode) Vector ===
+func derivImpl(timestamp clientmodel.Timestamp, args []Node) interface{} {
+	matrixNode := args[0].(MatrixNode)
+	resultVector := Vector{}
+
+	matrixValue := matrixNode.Eval(timestamp)
+	for _, samples := range matrixValue {
+		// No sense in trying to compute a derivative without at least two points.
+		// Drop this vector element.
+		if len(samples.Values) < 2 {
+			continue
+		}
+
+		// Least squares.
+		n := clientmodel.SampleValue(0)
+		sum_y := clientmodel.SampleValue(0)
+		sum_x := clientmodel.SampleValue(0)
+		sum_xy := clientmodel.SampleValue(0)
+		sum_x2 := clientmodel.SampleValue(0)
+		for _, sample := range samples.Values {
+			x := clientmodel.SampleValue(sample.Timestamp.UnixNano() / 1e9)
+			n += 1.0
+			sum_y += sample.Value
+			sum_x += x
+			sum_xy += x * sample.Value
+			sum_x2 += x * x
+		}
+		numerator := sum_xy - sum_x*sum_y/n
+		denominator := sum_x2 - (sum_x*sum_x)/n
+
+		resultValue := numerator / denominator
```

Comment:
It'll be +Inf

I had asked our stats guys about the cases where this would come up, and they didn't know. This seems like reasonable behaviour to me.